### PR TITLE
Fix attention application in DELG

### LINKED ISSUE #9877: DELG: Discrepancy on how local attention is computed
## 1. The entire URL of the file you are using

https://github.com/tensorflow/models/blob/master/research/delf/delf/python/training/model/delg_model.py

## 2. Describe the bug

In the DELG paper, attention is computed on block3, then applied to the output of the autoencoder. However, in the code, attention is computed from the output of the autoencoder. 

https://github.com/tensorflow/models/blob/417aa07384824ab567bd651444f5210e9f2e5bfb/research/delf/delf/python/training/model/delf_model.py#L210-L212

The attention model is only able to apply attention the the inputs provided - essentially self attention rather than cross attention used in the paper.

https://github.com/tensorflow/models/blob/417aa07384824ab567bd651444f5210e9f2e5bfb/research/delf/delf/python/training/model/delf_model.py#L68-L80

## 3. Steps to reproduce

Review this diagram

截屏2020-09-02 下午7 23 51

## 4. Expected behavior

Attention module should separately accept separate attention input and attention target, which would be `block3` and `dim_expanded_features` respectively for DELG.

This would allow an application of attention something like
`attn_prelogits, attn_scores, _ = self.attention(inputs=block3, targets=dim_expanded_features, training=training)`

### PR DESCRIPTION
**Pull request by its-dron**
# Description

> :memo: Please include a summary of the change. 
>  
> * Fixes #9877
> * DELG model applies attention to block 3 as in the paper, rather than to auto-encoder output, as in DELF.  
> * Enabled AttentionModel to specify an attention target in addition to attention source.

## Type of change

For a new feature or function, please create an issue first to discuss it
with us before submitting a pull request.

Note: Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)

## Tests

> :memo: Please describe the tests that you ran to verify your changes.
>  
> * Provide instructions so we can reproduce.  
> * Please also list any relevant details for your test configuration.  

**Test Configuration**:

## Checklist

- [x] I have signed the [Contributor License Agreement](https://github.com/tensorflow/models/wiki/Contributor-License-Agreements).
- [x] I have read [guidelines for pull request](https://github.com/tensorflow/models/wiki/Submitting-a-pull-request).
- [x] My code follows the [coding guidelines](https://github.com/tensorflow/models/wiki/Coding-guidelines).
- [x] I have performed a self [code review](https://github.com/tensorflow/models/wiki/Code-review) of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective or that my feature works.

### CODE REVIEW

**Comment by its-dron** [comment]
I don't have the resources to do a full training run to verify the affects of this change. 
Also note that the existing pretrained DELG models won't be affected, although they will still be using the incorrect attention.

File: research/delf/delf/python/training/model/delg_model_test.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+  )
+  def test_train_step(self, block3_strides):
+
+    image_size = 321
+    num_classes = 1000
+    batch_size = 2
+    clip_val = 10.0
+    input_shape = (batch_size, image_size, image_size, 3)
+
+    model = delg_model.Delg(block3_strides=block3_strides, name='DELG', use_dim_reduction=True)
```
Comment (andrefaraujo) [inline research/delf/delf/python/training/model/delg_model_test.py:110]: same comment about `name`

File: research/delf/delf/python/training/model/delg_model_test.py [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
+    loss_object = tf.keras.losses.SparseCategoricalCrossentropy(
+        from_logits=True, reduction=tf.keras.losses.Reduction.NONE)
+
+    def compute_loss(labels, predictions):
+      per_example_loss = loss_object(labels, predictions)
+      return tf.nn.compute_average_loss(
+          per_example_loss, global_batch_size=batch_size)
+
+    #TODO: use all 3 DELG losses here
```
Comment (andrefaraujo) [inline research/delf/delf/python/training/model/delg_model_test.py:129]: I think this comment can be removed since you added all 3 losses, right?

**Review by andrefaraujo — CHANGES_REQUESTED** [review]
Thanks Daniel! This looks great overall, happy to see this being fixed :)

Please take a look at the comments I left, most are small formatting issues (which are very common for those submitting PRs for the first time in this repo).

BTW, just one comment on what you mentioned in your previous message:
> Also note that the existing pretrained DELG models won't be affected, although they will still be using the incorrect attention

--> Actually, the pretrained models that are released in this repo are doing the correct thing, since they were actually trained using a different codebase.

**Comment by andrefaraujo** [comment]
@its-dron , friendly ping -- did you get time to update this? :)

File: research/delf/delf/python/training/model/delg_model_test.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+    model.init_classifiers(num_classes)
+
+    images = tf.random.uniform(input_shape, minval=-1.0, maxval=1.0, seed=0)
+    labels = tf.random.uniform((batch_size,),
+                               minval=0,
+                               maxval=model.num_classes - 1,
+                               dtype=tf.int64)
+    blocks = {}
+
+    # Get global feature by pooling block4 features.
```
[the remark this answers is not in the document]
Comment (danron-amazon) [reply research/delf/delf/python/training/model/delg_model_test.py:85]: I'm fairly certain both DELG and DELF use pooled block4 features for their global representation. This comment (and most of this code, for that matter) was copied from `delf_model_test.py`. 

We can also check the backbone code itself: 
```
if self.average_pooling:
  x = self.avg_pool(x)
  if intermediates_dict is not None:
    intermediates_dict['block4'] = x
else:
  if intermediates_dict is not None:
    intermediates_dict['block4'] = x

if self.include_top:
  return self.fc1000(self.flatten(x))
elif self.global_pooling:
  x = self.global_pooling(x)
  if self.embedding_layer:
    x = self.embedding_layer(x)
  return x
else:
  return x
```

File: research/delf/delf/python/training/model/delf_model.py [hunk 4]
```diff
@@ -209,6 +214,7 @@ def global_and_local_forward_pass(self, images, training=True):
     if self._use_dim_reduction:
       (dim_expanded_features, dim_reduced_features) = self.autoencoder(block3)
       attn_prelogits, attn_scores, _ = self.attention(dim_expanded_features,
+                                                      targets=block3,
```
[the remark this answers is not in the document]
Comment (danron-amazon) [reply research/delf/delf/python/training/model/delf_model.py:217]: Good catch!

**Comment by its-dron** [comment]
Apologies from the comments from my work account!

File: research/delf/delf/python/training/model/delg_model_test.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+    input_shape = (batch_size, image_size, image_size, 3)
+    local_feature_dim = 64
+    feature_map_size = image_size // 16 # reduction factor for resnet50.
+    if block3_strides:
+        feature_map_size //= 2
+
+    model = delg_model.Delg(block3_strides=block3_strides,
+                            use_dim_reduction=True,
+                            reduced_dimension=local_feature_dim
+                            )
```
Comment (andrefaraujo) [inline research/delf/delf/python/training/model/delg_model_test.py:48]: minor formatting nit: can you please move this close-parentheses to the previous line? Usually, per convention we leave the close-parentheses on the next line only if the previous one ends with a comma -- eg, replacing `=local_feature_dim` by `=local_feature_dim,` would also be fine. But maybe to keep conventions in these files, let's make it `=local_feature_dim)`

File: research/delf/delf/python/training/model/delg_model_test.py [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+    images = tf.random.uniform(input_shape, minval=-1.0, maxval=1.0, seed=0)
+
+    # Run a complete forward pass of the model.
+    global_feature, attn_scores, local_features = model.build_call(images)
+
+    self.assertAllEqual(global_feature.shape, (batch_size, 2048))
+    self.assertAllEqual(
+        attn_scores.shape,
+        (batch_size, feature_map_size, feature_map_size, 1)
+    )
```
Comment (andrefaraujo) [inline research/delf/delf/python/training/model/delg_model_test.py:60]: same

File: research/delf/delf/python/training/model/delg_model_test.py [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+    self.assertAllEqual(
+        local_features.shape,
+        (batch_size, feature_map_size, feature_map_size, local_feature_dim)
+    )
```
Comment (andrefaraujo) [inline research/delf/delf/python/training/model/delg_model_test.py:64]: same

File: research/delf/delf/python/training/model/delg_model_test.py [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+  def test_build_model(self, block3_strides):
+    image_size = 321
+    num_classes = 1000
+    batch_size = 2
+    input_shape = (batch_size, image_size, image_size, 3)
+
+    model = delg_model.Delg(
+        block3_strides=block3_strides,
+        use_dim_reduction=True
+    )
```
Comment (andrefaraujo) [inline research/delf/delf/python/training/model/delg_model_test.py:79]: same

File: research/delf/delf/python/training/model/delg_model_test.py [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+    image_size = 321
+    num_classes = 1000
+    batch_size = 2
+    clip_val = 10.0
+    input_shape = (batch_size, image_size, image_size, 3)
+
+    model = delg_model.Delg(
+        block3_strides=block3_strides,
+        use_dim_reduction=True
+    )
```
Comment (andrefaraujo) [inline research/delf/delf/python/training/model/delg_model_test.py:117]: same

File: research/delf/delf/python/training/model/delg_model_test.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (andrefaraujo) [reply research/delf/delf/python/training/model/delg_model_test.py:89]: for the DELG model, after pooling we add a projection (the line `x = self.embedding_layer(x)`), so the global feature is not directly the pooled block4 feature any more. See that the embedding layer is enabled here: https://github.com/tensorflow/models/blob/238922e98dd0e8254b5c0921b241a1f5a151782f/research/delf/delf/python/training/model/delg_model.py#L77

In the DELF case, that embedding layer is not used by default (so it is not used in the delf_model_test case).

In any case, this is just a small comment here :)  Maybe we can simply remove the comment line.

**Review by andrefaraujo — CHANGES_REQUESTED** [review]
Thanks again! Just a few more nits to fix, and I can immediately approve/merge once you make the changes :)

File: research/delf/delf/python/training/model/delg_model_test.py (hunk 6 above)
Comment (its-dron) [reply research/delf/delf/python/training/model/delg_model_test.py:60]: Quick check - is this indentation format correct? I see both this style and "indent to match open-paren" in the code.

Comment (andrefaraujo) [reply research/delf/delf/python/training/model/delg_model_test.py:60]: Actually, the easiest would be to just apply https://github.com/tensorflow/models/blob/master/.github/scripts/pylint.sh to the code. It should fix things automatically / point out other things that should be fixed as needed. Could you please try that?

File: research/delf/delf/python/training/model/delg_model_test.py (hunk 8 above)
Comment (andrefaraujo) [reply research/delf/delf/python/training/model/delg_model_test.py:75]: missing this one? :)

File: research/delf/delf/python/training/model/delg_model_test.py [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+    attn_logits = model.attn_classification(attn_prelogits)
+    self.assertAllEqual(attn_prelogits.shape, (batch_size, 1024))
+    self.assertAllEqual(attn_logits.shape, (batch_size, num_classes))
+
+  @parameterized.named_parameters(
+      ('block3_stridesTrue', True),
+      ('block3_stridesFalse', False),
+  )
+  def test_train_step(self, block3_strides):
+
```
Comment (andrefaraujo) [inline research/delf/delf/python/training/model/delg_model_test.py:102]: you can remove this empty line.

**Review by andrefaraujo — APPROVED** [review]
Thanks a lot!

**Pull request merged**